src/backend/functions/projects-edit.ts

```diff
@@ -0,0 +1,528 @@
+      originalProject.slideDeckLink,
+      body.slideDeckLink,
+      crId,
+      userId,
+      wbsElementId
+    );
+    if (slideChangeJson !== undefined) {
+      changes.push(slideChangeJson);
+    }
+  }
+  if (body.hasOwnProperty('bomLink')) {
+    const bomChangeJson = createChangeJsonNonList(
+      'bom link',
+      originalProject.bomLink,
+      body.bomLink,
+      crId,
+      userId,
+      wbsElementId
+    );
+    if (bomChangeJson !== undefined) {
+      changes.push(bomChangeJson);
+    }
+  }
+  if (body.hasOwnProperty('taskListLink')) {
+    const taskChangeJson = createChangeJsonNonList(
+      'task list link',
+      originalProject.taskListLink,
+      body.taskListLink,
+      crId,
+      userId,
+      wbsElementId
+    );
+    if (taskChangeJson !== undefined) {
+      changes.push(taskChangeJson);
+    }
+  }
+  if (body.hasOwnProperty('projectManager')) {
+    const projectManagerChangeJson = createChangeJsonNonList(
+      'project manager',
+      originalProject.wbsElement.projectManagerId,
+      body.projectManager,
+      crId,
+      userId,
+      wbsElementId
+    );
+    if (projectManagerChangeJson !== undefined) {
+      changes.push(projectManagerChangeJson);
+    }
+  }
+  if (body.hasOwnProperty('projectLead')) {
+    const projectLeadChangeJson = createChangeJsonNonList(
+      'project lead',
+      originalProject.wbsElement.projectLeadId,
+      body.projectLead,
+      crId,
+      userId,
+      wbsElementId
+    );
+    if (projectLeadChangeJson !== undefined) {
+      changes.push(projectLeadChangeJson);
+    }
+  }
+
+  // Dealing with lists
+  const rulesChangeJson = createRulesChangesJson(
+    'rules',
+    originalProject.rules,
+    rules,
+    crId,
+    userId,
+    wbsElementId
+  );
+  const goalsChangeJson = createDescriptionBulletChangesJson(
+    originalProject.goals.map((element) => descBulletConverter(element)),
+    goals,
+    crId,
+    userId,
+    wbsElementId,
+    'goals'
+  );
+  const featuresChangeJson = createDescriptionBulletChangesJson(
+    originalProject.features.map((element) => descBulletConverter(element)),
+    features,
+    crId,
+    userId,
+    wbsElementId,
+    'features'
+  );
+  const otherConstraintsChangeJson = createDescriptionBulletChangesJson(
+    originalProject.otherConstraints.map((element) => descBulletConverter(element)),
+    otherConstraints,
+    crId,
+    userId,
+    wbsElementId,
+    'other constraints'
+  );
+  // add the changes for each of dependencies, expected activities, and deliverables
+  changes = changes
+    .concat(rulesChangeJson)
+    .concat(goalsChangeJson.changes)
+    .concat(featuresChangeJson.changes)
+    .concat(otherConstraintsChangeJson.changes);
+
+  // update the project with the input fields
+  const updatedProject = await prisma.project.update({
+    where: {
+      wbsElementId
+    },
+    data: {
+      budget,
+      summary,
+      googleDriveFolderLink: body.googleDriveFolderLink,
```

Comment:
may as well just extract all of these at the beginning too. if they're undefined that's fine because it'll get set to undefined here

Reply:
That creates some weird bugs. I tried it and the code just became more complicated.